Fix wrong version set in TabletMeta and rowset null pointer

chaoyli opened the pull request.

yiguolei committed:
Remove alter task related header lock (#906)

yiguolei commented on be/src/olap/rowset/alpha_rowset_writer.cpp:
```diff
@@ -154,7 +154,7 @@ RowsetSharedPtr AlphaRowsetWriter::build() {
     for (auto& segment_group : _segment_groups) {
-        RETURN_NOT_OK(segment_group->load());
+        segment_group->load();
         bool checked = segment_group->check();
```
why not check segment group.load?

Commit:
Fix wrong version set in TabletMeta and rowset null pointer

1. Check whether rowset is nullptr or not when call add_unused_rowset()
2. Replace TabletMetaSharedPtr to eliminate wrong use

Commit:
Merge branch 'be_refactor' into be_refactor

kangpinghuang commented on be/src/olap/push_handler.cpp:
```diff
@@ -241,175 +241,174 @@ void PushHandler::_get_tablet_infos(const vector<TabletVars>& tablet_vars,
-  }
+    // open file
+    if (OLAP_SUCCESS != open(path, "rb")) {
+        OLAP_LOG_WARNING("fail to open file. [file='%s']", path);
```
LOG(WARNING)

kangpinghuang commented on be/src/olap/push_handler.cpp:
```diff
@@ -241,175 +241,174 @@ void PushHandler::_get_tablet_infos(const vector<TabletVars>& tablet_vars,
-  }
+    // load header
+    if (OLAP_SUCCESS != _header.unserialize(this)) {
+        OLAP_LOG_WARNING("fail to read file header. [file='%s']", path);
```
LOG(WARNING)

kangpinghuang commented on be/src/olap/push_handler.cpp:
```diff
@@ -241,175 +241,174 @@ void PushHandler::_get_tablet_infos(const vector<TabletVars>& tablet_vars,
+            if (_request.__isset.need_decompress && _request.need_decompress) {
+                need_decompress = true;
+            }
+            if (NULL == (reader = IBinaryReader::create(need_decompress))) {
```
(reader = IBinaryReader::create(need_decompress)) == nullptr

kangpinghuang commented on be/src/olap/snapshot_manager.cpp:
```diff
@@ -279,7 +279,7 @@ string SnapshotManager::_get_header_full_path(
 
 void SnapshotManager::update_header_file_info(
         const vector<RowsetSharedPtr>& consistent_rowsets,
-        TabletMeta* tablet_meta) {
```
why change pointer to shared_ptr?

kangpinghuang commented on be/src/olap/tablet.cpp:
```diff
@@ -177,19 +173,19 @@ OLAPStatus Tablet::revise_tablet_meta(const TabletMeta& tablet_meta,
         // load new local tablet_meta to operate on
-        TabletMeta new_tablet_meta;
-        TabletMetaManager::get_header(_data_dir, tablet_id(), schema_hash(), &new_tablet_meta);
+        TabletMetaSharedPtr new_tablet_meta(new TabletMeta());
```
TabletMetaSharedPtr new_tablet_meta(new(std::nothrow) TabletMeta())

kangpinghuang commented on be/src/olap/tablet.h:
```diff
@@ -214,7 +214,7 @@ class Tablet : public std::enable_shared_from_this<Tablet> {
 private:
     TabletState _state;
-    TabletMeta* _tablet_meta;
+    std::shared_ptr<TabletMeta> _tablet_meta;
```
TabletMetaSharedPtr

kangpinghuang commented on be/src/olap/tablet_manager.cpp:
```diff
@@ -686,16 +686,15 @@ TabletSharedPtr TabletManager::find_best_tablet_to_compaction(CompactionType com
 OLAPStatus TabletManager::load_tablet_from_meta(DataDir* data_dir, TTabletId tablet_id,
         TSchemaHash schema_hash, const std::string& meta_binary) {
-    std::unique_ptr<TabletMeta> tablet_meta(new TabletMeta());
+    TabletMetaSharedPtr tablet_meta(new TabletMeta());
```
TabletMetaSharedPtr tablet_meta(new(std::nothrow) TabletMeta());

kangpinghuang commented on be/src/olap/tablet_meta.cpp:
```diff
@@ -54,11 +54,11 @@ OLAPStatus TabletMeta::create(int64_t table_id, int64_t partition_id,
-                                  shard_id, tablet_schema,
-                                  next_unique_id, col_ordinal_to_unique_id);
+                              TabletMetaSharedPtr* tablet_meta) {
+    tablet_meta->reset(new TabletMeta(table_id, partition_id,
```
std::nothrow

kangpinghuang commented on be/src/olap/task/engine_storage_migration_task.cpp:
```diff
@@ -127,8 +127,8 @@ OLAPStatus EngineStorageMigrationTask::_storage_medium_migrate(
-        if (new_tablet_meta == NULL) {
+        TabletMetaSharedPtr new_tablet_meta(new(std::nothrow) TabletMeta());
+        if (new_tablet_meta == nullptr) {
             OLAP_LOG_WARNING("new olap header failed");
```
LOG(WARNING)

kangpinghuang commented on gensrc/proto/olap_file.proto:
```diff
@@ -270,7 +270,7 @@ message TabletMetaPB {
     optional TabletSchemaPB schema = 9;
     repeated RowsetMetaPB rs_metas = 10;
     repeated RowsetMetaPB inc_rs_metas = 11;
-    optional AlterTabletPB alter_tablet_task = 12;
```
why change this name?

kangpinghuang commented on be/src/olap/push_handler.cpp:
```diff
@@ -241,175 +241,174 @@ void PushHandler::_get_tablet_infos(const vector<TabletVars>& tablet_vars,
-                         << ", tablet=" << curr_tablet->full_name()
-                         << " read_rows=" << num_rows;
+        // 2. init RowsetBuilder of cur_tablet for current push
+        VLOG(3) << "init RowsetBuilder.";
```
RowsetWriter
